app/components/ui/molecules/SearchBox.test.js

```diff
@@ -0,0 +1,85 @@
+import React from 'react'
+import { mount } from 'enzyme'
+import { ThemeProvider } from 'styled-components'
+import { escapeRegExp } from 'lodash'
+import theme from '@elifesciences/elife-theme'
+
+import SearchBox from './SearchBox'
+
+const searchOptions = [
+  { value: 'first option' },
+  { value: 'final option' },
+  { value: 'second option' },
+]
+
+function filterFunction(options, searchValue, field) {
+  if (!searchValue) return options
+
+  const inputValue = searchValue.trim().toLowerCase()
+  if (!inputValue) return options
+
+  return options.filter(option =>
+    option[field].toLowerCase().includes(inputValue),
+  )
+}
+
+function getMatchIndex(inputValue, option) {
+  const re = new RegExp(escapeRegExp(inputValue))
+  const match = re.exec(option.toLowerCase())
+  if (match) return match.index
+  return -1
+}
+
+function makeWrapper() {
+  return mount(
+    <ThemeProvider theme={theme}>
+      <SearchBox
+        filterFunction={filterFunction}
+        getMatchIndex={getMatchIndex}
+        onSubmit={jest.fn()}
+        options={searchOptions}
+      />
+    </ThemeProvider>,
+  )
+}
+
+const wrapper = makeWrapper()
+
+describe('SearchBox component tests', () => {
+  it('renders an input field', () => {
+    expect(wrapper.find('input')).toHaveLength(1)
+  })
+
+  it('has search icon', () => {
+    expect(wrapper.find('#search-icon')).toHaveLength(1)
+  })
+
+  it("doesn't show any suggestions on empty input", () => {
+    wrapper.find('input').simulate('change', { target: { value: '' } })
+    expect(wrapper.find(SearchBox).instance().state.suggestions).toEqual([])
+  })
+
+  function compareSuggestions(searchValue) {
+    const expectedSuggestions = filterFunction(
+      searchOptions,
+      searchValue,
+      'value',
+    )
+    wrapper.find('input').simulate('change', { target: { value: searchValue } })
+    const receivedSuggestions = wrapper.find(SearchBox).instance().state
+      .suggestions
+    expect(receivedSuggestions).toEqual(expectedSuggestions)
+  }
+
+  it('typing something will show suggestions based on filtering function, test 1', () => {
+    compareSuggestions('option')
+  })
+
+  it('typing something will show suggestions based on filtering function, test 2', () => {
```

Comment:
Is there any added value in running this test 3 times?

Reply:
Yes, in retrospect I'm thinking I wrote too few tests for this :) What if there is a bug in the component so that when there's a match, only 2 results are suggested all the time? What if there is a bug where the filtering function is applied in a wrong way and it only matches first words? My take is that the fewer assumptions you make about the component you're testing, the better.

Reply:
Hmm, I'm still not sure that these are actually testing different pieces of the code.

First you simulate a change event on the input field. That's going to fire [the `onChange` handler in `react-autosuggest`](https://github.com/moroshko/react-autosuggest/blob/master/src/Autosuggest.js#L530-L549), which then calls the `onSuggestionsFetchRequested` prop that we've passed in.

I guess we're not trying to test whether the 3rd party component functions as it should. So these tests are all checking whether
- we have configured Autosuggest correctly in the render function (i.e. we have `onSuggestionsFetchRequested={this.onSuggestionsFetchRequested}`)
- we've written [our `onSuggestionsFetchRequested` function](https://github.com/elifesciences/elife-xpub/blob/c9cbd8a8dd633a7eb79d14d30e9e827392eb9e10/app/components/ui/molecules/SearchBox.js#L96-L110) correctly

More specifically, it's testing whether
- we have first called the `filterFunction` inside `onSuggestionsFetchRequested`
- whether state has then been updated with the return value

The only way any of these 3 tests would fail, is if we hadn't done the above correctly. So unless I'm missing something, I can't see a situation in which some of these 3 tests would pass and some would fail?

Reply:
That being said, I do agree that we need more tests in general, as we're not covering all of the functions in the `SearchBox` component. But I imagine that the bugs you're talking about are probably best caught by other tests?

For example, the following bug would be caught by testing the filterFunction itself:
> What if there is a bug where the filtering function is applied in a wrong way and it only matches first words?

As for this one
> What if there is a bug in the component so that when there's a match, only 2 results are suggested all the time?

that could be introduced by any number of things being implemented incorrectly:
- the render function in `SearchBox`
- the `renderSuggestions` function in `SearchBox`
- the onSuggestionsFetchRequested function in `SearchBox`
- the filterFunction in `PeoplePicker`

But! If there is an issue is with `onSuggestionsFetchRequested`, then all 3 tests will fail. If there is an issue in another place in the code, then all 3 tests will still fail. Right?

So, to find the cause of that bug, we would need tests for each of these functions.

Reply:
The way I'm testing things is more of black box testing[1]. Relying on implementation details when testing will cause you to catch less bugs. This is also the reason why testers are part of a different team than the developers most of the time. When I test I don't care about `react-autosuggest`, `onChange` handler or any of that. All I know is that I have a component called `SearchBox` and I know how it should behave and maybe how it should look like in the browser. Then I write tests to validate that behavior and look. If you want to write tests based on implementation details then be my guest (not that I approve of it), but I'm happy with the tests I've written.

[1] - https://en.wikipedia.org/wiki/Black-box_testing